File: Problem_2_Rotated_Sorted_array.java
class Solution {
  /*
  1) initializa low and high
  2) start binary search
        2.a) check if left is sorted
        2.b) check if right is sorted
        
        
      */
    public int search(int[] a, int target) {
      // 1
        int low = 0;
      int high = a.length -1;
      
      //2
      while(low <= high){
        
        int mid = low + (high - low)/2;
        
        //2.a
        if(a[low] == target) 
          return low;
        if(a[high] == target) 
          return high;
        if(a[mid] == target) 
          return mid;
        
        if(a[low] <= a[mid]){
          // left is sorted,
          if(a[low] < target && a[mid] > target){ // a[low] < target < a[mid] (does taeget fall in the sorted range?) if yes, do BS in left(sorted) range, if not, in right range
            high = mid -1;
          }else{
            low = mid + 1;
          }
        }else{
          //right is sorted
          if(a[high] > target && a[mid] < target ){ // a[mid] > target > a[high] (does taeget fall in the sorted range?) if yes, do BS in right(sorted) range, if not, in left range
            low = mid +1;
          }else{
            high = mid -1;
          }
        }
        
      }

        return -1;
      
        
    }
}

File: Problem-2_Rotated_Sorted_Array.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        #Initializing low and high index values of the array
        low = 0;
        high = len(nums)

        #Looping over until low index is less than high index value
        while low< high:
            #Calculating the mid index
            mid = (low + high) //2

            #If value at mid index equal to target return value at mid index
            if nums[mid] == target:
                return mid
            #If target value between low and mid index update low and high
            elif nums[low] <= nums[mid]:
                if target >= nums[low] and target <= nums[mid]:
                    high = mid
                else:
                    low = mid+1
            #If target value between mid and high index update low and high
            else:
                if target >= nums[mid] and target <=nums[high-1]:
                    low = mid +1
                else:
                    high = mid
        return -1


#Implementing the rotated sorted array using binary search with time complexity O(logn)
